crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -2718,6 +2791,76 @@ impl<'db> TypeInferenceBuilder<'db> {
         }
     }
 
+    /// Performs lexicographic comparison between two slices of types.
+    ///
+    /// This comparison supports only equality and ordering operators: `==`, `!=`, `<`, `<=`, `>`, `>=`.
+    /// For lexicographic comparison, elements from both slices are compared pairwise using
+    /// `infer_binary_type_comparison`. If a conclusive result cannot be determined as a `BoolLiteral`,
+    /// it returns `bool`.
+    fn infer_lexicographic_type_comparison(
+        &mut self,
+        left: &[Type<'db>],
+        op: ast::CmpOp,
+        right: &[Type<'db>],
+    ) -> Option<Type<'db>> {
+        if !matches!(
+            op,
+            ast::CmpOp::Eq
+                | ast::CmpOp::NotEq
+                | ast::CmpOp::Lt
+                | ast::CmpOp::LtE
+                | ast::CmpOp::Gt
+                | ast::CmpOp::GtE
+        ) {
+            debug_assert!(
+                false,
+                "Unsupported comparison operator for lexicographic type comparison."
+            );
+            return None;
+        }
+
+        // Compare paired elements from left and right slices
+        for (l_ty, r_ty) in left.iter().zip(right.iter()) {
```

Comment:
if you do this then you don't have to dereference the types very time you use them in the loop body below:

```suggestion
        for (l_ty, r_ty) in left.iter().copied().zip(right.iter().copied()) {
```